src/sourmash/tax/tax_utils.py

```diff
@@ -36,7 +39,272 @@
+
+        return zipped
+
+    def display_lineage(self, truncate_empty=True, null_as_unclassified=False):
+        "Return lineage names as ';'-separated list"
+        lin = ";".join(self.zip_lineage(truncate_empty=truncate_empty))
+        if null_as_unclassified:
+            if lin == "":
+                return "unclassified"
+        else:
+            return lin
+
+    def display_taxid(self, truncate_empty=True):
+        "Return lineage taxids as ';'-separated list"
+        return ";".join(self.zip_taxid(truncate_empty=truncate_empty))
+
+    def check_rank_availability(self, rank):
+        if rank in self.ranks: # rank is available
+            return True
+        raise ValueError(f"Desired Rank '{rank}' not available for this lineage.")
+ 
+    def rank_is_filled(self, rank, other=None):
+        self.check_rank_availability(rank)
+        if other is not None:
+            if rank in self.filled_ranks and rank in other.filled_ranks:
+                return True
+        elif rank in self.filled_ranks:
+            return True
+        return False
+
+    def is_lineage_match(self, other, rank):
+        """
+        check to see if two lineages are a match down to given rank.
+        """
+        if not other.ranks == self.ranks: # check same ranks
+            raise ValueError("Cannot compare lineages from taxonomies with different ranks.")
+        # always return false if rank is not filled in either of the two lineages
+        if self.rank_is_filled(rank, other=other):
+            rank_idx = self.rank_index(rank)
+            a_lin = self.lineage[:rank_idx+1]
+            b_lin = other.lineage[:rank_idx+1]
+            if a_lin == b_lin:
+                return 1
+        return 0
+
+    def pop_to_rank(self, rank):
+        "Return new LineageInfo with ranks only filled to desired rank"
+        # are we already above rank?
+        if not self.rank_is_filled(rank):
+            return replace(self)
+        # if not, make filled_lineage at this rank + use to generate new LineageInfo
+        new_lineage = self.lineage_at_rank(rank)
+        new = replace(self, lineage = new_lineage)
+        # replace doesn't run the __post_init__ properly. reinitialize.
+        new._init_from_lineage_tuples()
+        return new
+
+    def lineage_at_rank(self, rank):
+        # non-descructive pop_to_rank. Returns tuple of LineagePairs
+        "Returns tuple of LineagePairs at given rank."
+        # are we already above rank?
+        if not self.rank_is_filled(rank):
+            return self.filled_lineage
+        # if not, return lineage tuples down to desired rank
+        rank_idx = self.rank_index(rank)
+        return self.filled_lineage[:rank_idx+1]
+
+
+@dataclass(frozen=True, order=True)
+class RankLineageInfo(BaseLineageInfo):
+    """Class for storing multi-rank lineage information"""
+    ranks: tuple = field(default_factory=lambda: ('superkingdom', 'phylum', 'class', 'order', 'family', 'genus', 'species', 'strain'))
```

Comment:
To return to later: is `field` a new keyword or something? does it only work in dataclasses?

Reply:
`field` is a keyword within dataclasses that is used to pass additional information.

The `field(default_factory=)` format is required for defaults that are iterable and mutable (e.g. lists). Now that `ranks` is a tuple (for hashability), this format is not actually required here - will fix.